File: javascript/lesson 2/detectCycle.js
function detectCycle(head) {
    const intersect = head && getIntersect(head);
    return intersect && getCycleStart(head, intersect);
}

function getIntersect(head) {
    let slow = head, fast = head;
    while (fast && fast.next) {
        slow = slow.next;
        fast = fast.next.next;
        if (slow === fast) return slow;
    }
    return null;
}

function getCycleStart(head, intersect) {
    let one = head, two = intersect;        
    while (one !== two) {
        one = one.next;
        two = two.next;
    }
    return one;
}

File: ruby/lesson 2/detect_cycle.rb
def detect_cycle(head)
    return nil if head == nil
    intersect = get_intersect(head)
    intersect && get_cycle_start(head, intersect)
end

def get_intersect(head)
    slow = fast = head
    while fast&.next
        slow = slow.next
        fast = fast.next.next
        return slow if slow == fast
    end
    nil
end

def get_cycle_start(head, intersect)
    one = head
    two = intersect
    while one != two
        one = one.next
        two = two.next
    end
    one
end
